### PR TITLE
[16.0][ADD] product_packaging_level_purchasable

### PR DESCRIPTION
**Pull request by santostelmo**
This module feature was extracted from the original [sale-workflow/sale_by_packaging](https://github.com/oca/sale-workflow/tree/14.0/sale_by_packaging) module.

This module allows odoo users to set:

Can be sold: On product packaging type model, this checkbox defines if product
packagings from this particular type are available to be selected on purchase
order line.

### CODE REVIEW

File: product_packaging_level_purchasable/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
+# Copyright 2023 Camptocamp SA
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl)
+{
+    "name": "Product Packaging level purchasable",
+    "summary": "",
+    "version": "16.0.1.0.0",
+    "development_status": "Alpha",
+    "category": "Product",
+    "website": "https://github.com/OCA/product-attribute",
+    "author": "Camptocamp, Odoo Community Association (OCA)",
```
Comment (jbaudoux) [inline product_packaging_level_purchasable/__manifest__.py]:
```suggestion
    "author": "Camptocamp, BCIM, Odoo Community Association (OCA)",
```

File: product_packaging_level_purchasable/models/product_packaging.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
+# Copyright 2023 Camptocamp SA
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl)
+
+from odoo import api, fields, models
+
+
+class ProductPackaging(models.Model):
+    _inherit = "product.packaging"
+
+    purchases = fields.Boolean(
```
Comment (jbaudoux) [inline product_packaging_level_purchasable/models/product_packaging.py]: The field name could be improved
```suggestion
    can_be_purchased = fields.Boolean(
```

File: product_packaging_level_purchasable/models/purchase_order_line.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+                )
+
+    @api.onchange("product_packaging_id")
+    def _onchange_product_packaging_id(self):
+        if self._can_be_purchased_error_condition():
+            return {
+                "warning": {
+                    "title": _("Warning"),
+                    "message": _(
+                        "This product packaging must be set as 'Purchases' in"
```
Comment (jbaudoux) [inline product_packaging_level_purchasable/models/purchase_order_line.py]:
```suggestion
                        "This product packaging must be set as 'Cab be purchased' in"
```

File: product_packaging_level_purchasable/readme/CONFIGURE.rst [hunk 4]
```diff
@@ -0,0 +1,5 @@
+This module allows odoo users to set:
+
+* Can be purchased: On product packaging level model, this checkbox defines if product
+  packagings from this particular type are available to be selected on purchase
+  order line.
```
Comment (jbaudoux) [inline product_packaging_level_purchasable/readme/CONFIGURE.rst]:
```suggestion
* Can be purchased: On a product packaging level as default value for a new product packaging. On a product packaging to make it available or not on a purchase order line.
```

**Review by jbaudoux — COMMENTED** [review]
LG. The boolean field name could be improved

**Comment by jbaudoux** [comment]
This is inspired from sales equivalent https://github.com/OCA/product-attribute/pull/1395

File: product_packaging_level_purchasable/models/product_packaging.py [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+        compute="_compute_purchases",
+        readonly=False,
+        store=True,
+        default=None,
```
Comment (simahawk) [inline product_packaging_level_purchasable/models/product_packaging.py]:
```suggestion
```

File: product_packaging_level_purchasable/models/purchase_order_line.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    _inherit = "purchase.order.line"
+
+    def _can_be_purchased_error_condition(self):
+        self.ensure_one()
+        return self.product_packaging_id and not self.product_packaging_id.purchases
+
+    @api.constrains("product_packaging_id")
+    def _check_product_packaging_can_be_purchased(self):
+        for line in self:
+            if line._can_be_purchased_error_condition():
```
Comment (simahawk) [inline product_packaging_level_purchasable/models/purchase_order_line.py:21]: pls collect all lines failing before throwing an error

**Review by simahawk — CHANGES_REQUESTED** [review]
pls address change requests

**Comment by santostelmo** [comment]
@simahawk @jbaudoux changes done

**Review by simahawk — APPROVED** [review]
LGTM

**Review by cyrilmanuel — APPROVED** [review]

**Pull request merged**